doc/port-profile-init/port-profile-init-design.md

```diff
@@ -0,0 +1,179 @@
+
+### Scope
+
+This document is the design document for Port Profile Init feature in SONiC.
+
+### Definitions/Abbreviations
+
+N/A
+
+### Overview
+
+Port configuration is currently done in 2 phases.
+* **Static phase:** SAI parses SAI profile file and creates all ports there one by one.
+* **Dynamic phase:** SONiC PortsOrch has a comparison process between the pre-created ports from SAI, and ports configurations in CONFIG DB. In case there is a mismatch with one of the ports, PortsOrch will remove the port and recreate it with right lanes and attributes.
+
+The above logic requires multiple calls to SAI APIs and can take significant amount of time, which can challenge fast-boot time restriction.
+
+SAI introduced new capability which allows using bulk operation for creating any type of object. Using these capability will allow us saving API calls and decrease ports creation time.
+
+### Requirements
+
+* SONiC will support both bulk approach and legacy approach.
+* Before creating the ports, SONiC will check if bulk operation is supported by SAI. If not, it will fallback to legacy approach.
+* Ports creation bulk approach in SONiC expectes SAI to not create the port in create_switch() phase.
+* Before creating the ports, even if bulk is suported, SONiC will check if SAI created ports as before. If so, it will fallback to legacy approach.
+* SONiC will send the bulk request only when all ports are configured in APP DB, PORT table. Meaning, only when "PortConfigDone" flag is raised.
+* In case create_ports() API is called with an empty list of ports, SONiC expects SAI to return a status different from SAI_STATUS_NOT_IMPLEMENTED or SAI_STATUS_NOT_SUPPORTED. This call will be made in order to check if bulk approach is supported and implemented in vendor's SAI.
+* After feature is implemented, SAI initialization time should be reduced. Hence, we expect fast-boot up time to meet max 30 seconds restriction.
+
+
+### Architecture Design
+
+This feature does not change the existing SONiC architecture, while it has to change the configuration flow for ports. This change will be covered in orchagent - PortsOrch.
+
+
+### High-Level Design
+
+There will be no changes in CLI, DB schema, YANG model or warm-reboot support.
+
+In order to support feature in SONiC side, Port configuration flow will be changed in sonic-swss PortsOrch module.
+
+<p align=left>
+<img src="img/old_flow.svg" alt="Figure 1. Old Flow">
+</p>
+
+
+<p align=left>
+<img src="img/new_flow.svg" alt="Figure 2. New Flow">
+</p>
+
+
+The above sequence diagram shows that a lot of calls to SAI and SDK APIs are being saved.
+In terms of Fast Reboot, when system is going up, it can save a lot of time and not challenge the 30 seconds restriction.
+
+Notes:
+
+* Multi-Asic is covered for these changes. In Multi-Asic environment, there are multiple instances of swss and Redis DB and each DB has it's own PORT table. Hence, port configuration is per Asic.
```

Comment:
Can you list what are the use cases need bulk API support? what is wrong with existing SAI create ports? Is that DPB needs bulk API support ?

Reply:
Bulk API is needed in order to save time calling the regular APIs one by one. 
It doesn't mean there is something wrong with the current approach, just that there is a faster way which will help the initialization to be faster and the plan is to save time for fast-boot so the uptime will not cross the 30 seconds limitation.